### PR TITLE
Add special test suite to probe all precondition combinations 

### PR DESCRIPTION
**Pull request by hegyibalint**
This PR adds the measurement part of the precondition framework.
To measure which preconditions are running and where this PR implements a special test suite.

The test suite is configured to being able to run with all `Test.class` based tasks.

### CODE REVIEW

File: subprojects/precondition-tester/README.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Gradle / Predicate Tester
```
Comment (wolfs) [inline subprojects/precondition-tester/README.md:1]: 🎉 Nice adding a README here!

File: subprojects/precondition-tester/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ */
+
+plugins {
+    id("gradlebuild.internal.java")
+}
+
+description = "Internal project testing and collecting information about all the test preconditions."
+
+// Special configuration containing all the precondition classes
+val precondition by configurations.creating {
```
Comment (wolfs) [inline subprojects/precondition-tester/build.gradle.kts:24]: 🤔 Instead of creating a new configuration, could we add the dependencies to `testFixtures` and that would work the same way?

File: subprojects/precondition-tester/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    description = "Configuration for all TestPrecondition classes dynamically loaded by the precondition tester"
+    isCanBeResolved = false
+    isCanBeConsumed = true
+}
+
+configurations {
+    // All test runtime configurations should extend from the "precondition" configuration
+    listOf(testRuntimeOnly, integTestRuntimeOnly, archTestRuntimeOnly, crossVersionTestRuntimeOnly).forEach {
```
Comment (wolfs) [inline subprojects/precondition-tester/build.gradle.kts:32]:
```suggestion
    listOf(testRuntimeOnly, integTestRuntimeOnly, crossVersionTestRuntimeOnly).forEach {
```

🤔 Why wire the tests into `archTestRuntimeOnly` as well? Those do not support preconditions, so they shouldn't run the precondition tests, right?

File: subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/LocalPreconditionProbingTests.groovy [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.test.precondition
+
+import com.gradle.enterprise.testing.annotations.LocalOnly
+
+@LocalOnly
+class LocalPreconditionProbingTests extends PreconditionProbingTests {
```
Comment (wolfs) [inline subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/LocalPreconditionProbingTests.groovy:22]:
```suggestion
class LocalPreconditionProbingTests extends PreconditionProbingTest {
```

File: subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ *         <b>Skipped:</b> the precondition combination could <i>not</i> be satisfied.
+ *         This is normal, and simply signifies that the precondition is not satisfied on the current system
+ *     </li>
+ *     <li>
+ *         <b>Failed:</b> there was an error meanwhile checking the combination.
+ *         This shouldn't be normal, and needs fixing.
+ *     </li>
+ * </ul>
+ */
+abstract class PreconditionProbingTests extends Specification {
```
Comment (wolfs) [inline subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy:40]:
```suggestion
abstract class PreconditionProbingTest extends Specification {
```

File: subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    Class<? extends TestPrecondition> loadClass(String className) {
+        println(System.getProperty("java.class.path"))
```
Comment (wolfs) [inline subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy:43]: ❌ Another debug output that should probably go away.

File: subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    def "Preconditions #classNames"() {
+        given:
+        def classes = classNames.collect { loadClass(it) }
+
+        expect:
+        for (def precondition : classes) {
+            def preconditionInstance = precondition.getDeclaredConstructor().newInstance()
+            Assumptions.assumeTrue(preconditionInstance.satisfied)
```
Comment (wolfs) [inline subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy:62]: 🤔 Can you use `TestPrecondition.doSatisfiesAll()` directly here?

**Review by wolfs — CHANGES_REQUESTED** [review]

File: subprojects/precondition-tester/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            enabled = false
+        }
+
+        // These tests should always run
+        outputs.upToDateWhen { false }
+    }
+}
+
+afterEvaluate {
+    configurations.forEach(::println)
```
[the remark this answers is not in the document]
Comment (hegyibalint) [reply subprojects/precondition-tester/build.gradle.kts:86]: Correct, was used to debug

File: subprojects/precondition-tester/build.gradle.kts (hunk 2 above)
Comment (hegyibalint) [reply subprojects/precondition-tester/build.gradle.kts:24]: ❓ Does `testFixtures` getting inherited to all the testing configurations?
I had a real issue trying to declare dependencies (in a tidy way) to all the testing configurations.

File: subprojects/precondition-tester/build.gradle.kts (hunk 3 above)
Comment (hegyibalint) [reply subprojects/precondition-tester/build.gradle.kts:32]: I wasn't thinking about that `archTest` doesn't really need the preconditions...
In this case, I guess, https://github.com/gradle/gradle/blob/52c36b9ca82047c9731da374f00fd6027983b65f/subprojects/precondition-tester/build.gradle.kts#L67-L83 needs some modifications also, so we skip executing when `arch...` tests are being done.

File: subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy (hunk 6 above)
Comment (hegyibalint) [reply subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy]: Removed

File: subprojects/precondition-tester/build.gradle.kts (hunk 3 above)
Comment (wolfs) [reply subprojects/precondition-tester/build.gradle.kts:32]: Yeah, let's hook this into the test tasks we know about instead of every test task we find.

File: subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/LocalPreconditionProbingTests.groovy (hunk 4 above)
Comment (hegyibalint) [reply subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/LocalPreconditionProbingTests.groovy]: Done

File: subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy (hunk 5 above)
Comment (hegyibalint) [reply subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTests.groovy]: Done

File: subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTest.groovy [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        def classes = classNames.collect { loadClass(it) }
+
+        expect:
+        for (def precondition : classes) {
+            def preconditionInstance = precondition.getDeclaredConstructor().newInstance()
+            Assumptions.assumeTrue(preconditionInstance.satisfied)
+        }
+
+        where:
+        classNames << PredicatesFile.DEFAULT_ACCEPTED_COMBINATIONS
```
[the remark this answers is not in the document]
Comment (hegyibalint) [reply subprojects/precondition-tester/src/test/groovy/org/gradle/test/precondition/PreconditionProbingTest.groovy:65]: The file is not generated, it's a file updated by the developers when they need a new combination.

Correctness is made certain by tests on both sides:
 - If regular tests with @Requires reference combinations that do not exist, tests are going to fail.
 - If the file is incorrect, the RequiresExtension is going to fail with an exception
 These cases are visible to the user and verify the integrity of the predicate combinations.

Additionally, this test will fail as `PredicatesFile` will return an exception if the file is incorrect.

TL;DR: tests will fail if the file is incorrect.

File: subprojects/precondition-tester/build.gradle.kts (hunk 3 above)
Comment (hegyibalint) [reply subprojects/precondition-tester/build.gradle.kts]: Currently worked around by not configuring `archTest`. 
Sadly it seems that the test doesn't have its own test class, so we have to resort to task name matching.

File: subprojects/precondition-tester/build.gradle.kts (hunk 2 above)
Comment (hegyibalint) [reply subprojects/precondition-tester/build.gradle.kts]: Actually worked out!

File: subprojects/precondition-tester/build.gradle.kts (hunk 3 above)
Comment (wolfs) [reply subprojects/precondition-tester/build.gradle.kts]: We could cover the regular `test` task and `DistributionTest`s. For the testing, we might want to check covered preconditions per task name, so we'd notice immediately when a new test task arises that we don't cover.

File: subprojects/precondition-tester/README.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+- The test is running on Windows (`@Requires(UnitTestPreconditions.Windows)`)
+- The test is running on JDK 8 compatible JDKs (`@Requires(UnitTestPreconditions.Jdk8OrLater)`)
+- Arbitrary combinations of above (`@Requires([UnitTestPreconditions.Windows, UnitTestPreconditions.Jdk8OrLater])`)
+
+The issue is that we _ignore_ tests, where the predicate is not satisfied.
+Albeit being the correct solution, this can cause problems where a test is _not running anywhere_ and we don't know anything about it.
+
+For example, a `RunsOnPDP11` predicate, which we won't satisfy anywhere as we don't run tests on such hardware, will always be ignored.
+
+Multiple occasions happened when engineers were caught by surprise finding tests not running anywhere (e.g., tests depending on non-LTS, cleaned up JDK distributions). This project is the result of an initiative to [redesign the predicate system](, where we aim to precisely track if all predicates are satisfied
```
[the remark this answers is not in the document]
Comment (hegyibalint) [reply subprojects/precondition-tester/README.md]: Fixed

File: subprojects/precondition-tester/build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+description = "Internal project testing and collecting information about all the test preconditions."
+
+dependencies {
+    // ========================================================================
+    // All subprojects, which has their own preconditions.
+    // These projects should have their preconditions in the "src/testFixtures" sourceSet
+    // ========================================================================
+    // This whole project is for test support, i.e. it's "main" source set is used
```
[the remark this answers is not in the document]
Comment (hegyibalint) [reply subprojects/precondition-tester/build.gradle.kts:28]: Added

File: subprojects/precondition-tester/build.gradle.kts [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    // Other, project-related dependencies
+    // ========================================================================
+    testImplementation(libs.junit5JupiterApi) {
+        because("Assume API comes from here")
+    }
+}
+
+tasks {
+    test {
+        this.setupAsProbingTest()
```
Comment (wolfs) [inline subprojects/precondition-tester/build.gradle.kts:66]:
```suggestion
        setupAsProbingTest()
```

You can drop the `this`, right?

File: subprojects/precondition-tester/build.gradle.kts [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    withType(DistributionTest::class) {
+        this.setupAsProbingTest()
+    }
+}
+
+fun Test.setupAsProbingTest() {
```
Comment (wolfs) [inline subprojects/precondition-tester/build.gradle.kts:74]:
```suggestion
fun Test.setupPreconditionTesting() {
```

Maybe that is better? We don't talk about `probing` anywhere, yet, right?

**Comment by hegyibalint** [comment]
@bot-gradle test this

File: subprojects/precondition-tester/build.gradle.kts (hunk 12 above)
Comment (hegyibalint) [reply subprojects/precondition-tester/build.gradle.kts:66]: Removed

**Comment by hegyibalint** [comment]
@bot-gradle test and merge

**Comment by hegyibalint** [comment]
@bot-gradle cancel

**Comment by hegyibalint** [comment]
@bot-gradle test and merge

**Comment by hegyibalint** [comment]
@bot-gradle test and merge
